deploy/operator.yaml

```diff
@@ -34,6 +34,9 @@ spec:
             fieldRef:
               fieldPath: metadata.name
         - name: WATCH_NAMESPACE
+          valueFrom:
```

Comment:
You can not do this. Our ability to get conditions from objects that are not in our namespace (ie. CDI, ClusterNetworkAddons, etc) relies on watching all namespaces.

Reply:
> You can not do this. Our ability to get conditions from objects that are not in our namespace (ie. CDI, ClusterNetworkAddons, etc) relies on watching all namespaces.

The rationale of this pull request was to grey-out the default installation mode (Operator will be available in all namespaces), so that it will be possible to install the operator only at the designated namespace, as required.
This method is implemented by many other operators (eg. 3scale, Prometheus Operator, Fuse Online Operator, etc.)
By checking the CSV of these operators, this is implemented by setting the installModes.type: AllNamespaces to supported: false.

However, that way I'm getting `AllNamespaces InstallModeType not supported, cannot configure to watch all namespaces` from OLM so I tried to limit it to one namespace.

CDI and CNAO are cluster-scoped objects, thus are not bound to any specific namespace. Do we need "AllNamespaces" watch permissions to get conditions from cluster-scoped resources?

Reply:
I would be very surprised if OLM were actually looking at what is put into the `WATCH_NAMESPACE` environment varaible for any/all containers in a CSV.

Looking at the source code for OLM, [this](https://github.com/operator-framework/operator-lifecycle-manager/blob/372a68604a028eb9c0d61ce7beda0bf948ba11a6/pkg/api/apis/operators/v1alpha1/clusterserviceversion.go#L178) is where the error you are seeing comes from and [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/5fd545c78a957c45d74651dab56192f00ae42b13/pkg/controller/operators/olm/operator.go#L1213-L1217) is where this function is called.

Let us look at this a little more closely. First how the function is called:

```
modeSet.Supports(out.GetNamespace(), namespaces)
```

It looks like `out.GetNamespace()` will return with the namespace where the CSV is installed (ie. where the Subscription is created). The value of `namespaces` is a slice of strings that is the result of splitting the result of this line `targets, ok := out.GetAnnotations()[v1.OperatorGroupTargetsAnnotationKey]`. `out` here is a CSV so we are looking for `olm.targetNamespaces` annotation on the CSV. Reading through [OLMs documentation on Operator Groups](https://github.com/operator-framework/operator-lifecycle-manager/blob/85375dd71d61b2db1a0a7173723dd97132b8c339/doc/design/operatorgroups.md#operatorgroup-csv-annotations) you will see that the value of the `olm.targetNamespaces` annotation comes from the spec of the OperatorGroup. 

**tl;dr**

The arguments given to the call to the `Supports(` function are:

1. The namespace where the CSV was installed (or, from a user perspective, where the Subscription was created)
1. The `spec.targetNamespaces` of the OperatorGroup

Now let us look at the `Supports(` function, trying to limit to the necessary bits:

```
func (set InstallModeSet) Supports(operatorNamespace string, namespaces []string) error {
	numNamespaces := len(namespaces)
	switch {
	case numNamespaces == 1:
		switch namespaces[0] {
		case v1.NamespaceAll:
			if !set[InstallModeTypeAllNamespaces] {
				return fmt.Errorf("%s InstallModeType not supported, cannot configure to watch all namespaces", InstallModeTypeAllNamespaces)
			}
		}
	}
```

Based on the code path we can determine that the `spec.TargetNamespaces` for the OperatorGroup was `*`.

**Solution:** You will need to be certain that the OperatorGroup, that you create or is created for you, has `spec.targetNamespaces` equivalent to the namespace where the Subscription is made.

Reply:
Hi David,
I already configured the OperatorGroup `spec.targetNamespaces` field with the proper namespace where the subscription is made ("kubevirt-hyperconverged" in our case):
`upgrade-test.sh`:
```
cat <<EOF | ${CMD} create -f -
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: hco-operatorgroup
  namespace: kubevirt-hyperconverged
  spec:
    targetNamespaces:
    - kubevirt-hyperconverged
EOF
```
https://github.com/orenc1/hyperconverged-cluster-operator/blob/master/hack/upgrade-test.sh#L118
But still, after subscription deployment I'm getting the following fail message:
`AllNamespaces InstallModeType not supported, cannot configure to watch all namespaces`

Is there something else I'm missing?

Reply:
> Is there something else I'm missing?

Nothing stands out as wrong about what you are doing. It may be appropriate to reach out to the OLM folks and see what they say. I don't see anything outside of the OperatorGroup that could cause this error message.